include/xtensor/xmasked_value.hpp

```diff
@@ -0,0 +1,817 @@
+                                                                                              \
+        template <class T1, class B1>                                                         \
+        inline xmasked_value& operator OP(const xtl::xoptional<T1, B1>& value)                \
+        {                                                                                     \
+            m_flag = m_flag && value.has_value();                                             \
+            if (m_flag)                                                                       \
+            {                                                                                 \
+                m_value OP value.value();                                                     \
+            }                                                                                 \
+            return *this;                                                                     \
+        }                                                                                     \
+                                                                                              \
+        template <class T1, class B1>                                                         \
+        inline xmasked_value& operator OP(const xmasked_value<T1, B1>& value)                 \
+        {                                                                                     \
+            m_flag = m_flag && value.has_value();                                             \
+            if (m_flag)                                                                       \
+            {                                                                                 \
+                m_value OP value.value();                                                     \
+            }                                                                                 \
+            return *this;                                                                     \
+        }
+
+        DEFINE_ASSIGN_OPERATOR(=);
+        DEFINE_ASSIGN_OPERATOR(+=);
+        DEFINE_ASSIGN_OPERATOR(-=);
+        DEFINE_ASSIGN_OPERATOR(*=);
+        DEFINE_ASSIGN_OPERATOR(/=);
+        DEFINE_ASSIGN_OPERATOR(%=);
+        DEFINE_ASSIGN_OPERATOR(&=);
+        DEFINE_ASSIGN_OPERATOR(|=);
+        DEFINE_ASSIGN_OPERATOR(^=);
+
+#undef DEFINE_ASSIGN_OPERATOR
+
+        inline void swap(xmasked_value& other)
+        {
+            std::swap(m_value, other.m_value);
+            std::swap(m_flag, other.m_flag);
+        }
+
+    private:
+
+        value_type m_value;
+        flag_type m_flag;
+    };
+
+    template <class T>
+    inline auto masked_value(T&& val) -> disable_xoptional_like<std::decay_t<T>, xmasked_value<T>>
+    {
+        return xmasked_value<T>(std::forward<T>(val));
+    }
+
+    template <class T, class B>
+    inline auto masked_value(const xtl::xoptional<T, B>& val)
```

Comment:
This is not required, we can build a `xmasked_value` whose `value_type` is an optional, the behavior should be the same until the final conversion (when assigning an `xmasked_view` to a variable of its `value_type`).

If you want to build an `xmasked_view` from an `xoptional` because the `xoptional` flag member will be used as the mask member of the `xmasked_value`, this should be done explicitly with the overload below.